### PR TITLE
Index spi: all types

### PR DESCRIPTION
**Pull request by gortiz**
This is a partial draft PR related to https://github.com/apache/pinot/issues/10183. The whole PR is in https://github.com/apache/pinot/pull/10184. It isn't intended to be merged as it is and it doesn't require to compile or pass the checks. It has been created in order to make it easier to understand the PR with all the changes.

This PR depends on https://github.com/apache/pinot/pull/10192. Readers only interested in the difference between this and the previous PR can focus on https://github.com/gortiz/pinot/pull/2

This PR shows all IndexTypes partially implemented without actually modifying the code.

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/bloom/BloomIndexType.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import org.apache.pinot.spi.config.table.BloomFilterConfig;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.Schema;
+
+public class BloomIndexType implements IndexType<BloomFilterConfig, BloomFilterReader, BloomFilterCreator> {
+  public static final BloomIndexType INSTANCE = new BloomIndexType();
+
+  @Override
+  public String getId() {
+    return "bloom_filter";
```
Comment (saurabhd336) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/bloom/BloomIndexType.java:47]: Could these Ids be added to a Constants file? Alteast for the ones defined in StandardIndexes.java?

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    }
+    return new IndexService(pluginList);
+  }
+
+  public Set<IndexType<?, ?, ?>> getAllIndexes() {
+    return _allIndexes;
+  }
+
+  public Optional<IndexType<?, ?, ?>> get(String indexId) {
+    return getAllIndexes().stream().filter(indexType -> indexType.getId().equalsIgnoreCase(indexId)).findAny();
```
Comment (saurabhd336) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java:76]: Could we maybe have a map of id -> IndexType so we don't have do this each time?

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ * Ideally, if we used some kind of injection system, this class should be injected into a Pinot context all classes can
+ * receive when they are built. Given that Pinot doesn't have that, we have to relay on static fields.
+ *
+ * By default, this class will be initialized by reading all ServiceLoader SPI services that implement
+ * {@link IndexPlugin}, adding all the {@link IndexType} that can be found in that way.
+ *
+ * In case we want to change the instance to be used at runtime, the method {@link #setInstance(IndexService)} can be
+ * called.
+ */
+public class IndexService {
```
Comment (saurabhd336) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java:41]: Is there a reason why an instance of this class need to exist? Could _allIndexes be initialised statically and `getOrThrow(String indexId)` be a static method?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/bloom/BloomIndexType.java (hunk 1 above)
Comment (gortiz) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/bloom/BloomIndexType.java:47]: I like the idea of adding them as constant in StandardIndexes and then consume the constants on each IndexType implementation.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java (hunk 2 above)
Comment (gortiz) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java:76]: yes, we can. Although this method is not required to be that fast. It is not in the hotpath or anything. But it is a simple change

Comment (gortiz) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java:76]: BTW, I think I'm going to do this change in index-spi-index-service branch instead

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java (hunk 3 above)
Comment (gortiz) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java:41]: We have been discussing about this in https://github.com/apache/pinot/pull/10192 and also in the PEP ([link](https://docs.google.com/document/d/1L_GJ2tAgZVxI_Ssa59--Mso3mK_HeAS7s7G9261-sOE/edit?disco=AAAAsjZUDEA)).

To have singletons or statics is something we don't like, but AFAIK is the only thing we can do right now. My suggestion is to have a context we can inject in places where we need these contextual information. Alternatively we could use ThreadLocal. The idea here is to have the logic in object methods instead that in class methods in order to be closer to the ideal world without statics.

Meanwhile, by having the instance, we can write the code in a way that it can be tested without static components. I mean, we can consume this class with `IndexService.getInstance().getAllIndexes()` or we can receive the instance as a construction argument and it an attribute to later do `_indexService.getAllIndexes()`. If we do the later, we can easily create test scenarios with different index type sets. Otherwise to create a test that do that we need to modify the static variable, which means that it is more difficult (if even possible) to run tests in parallel.

I have to admit that the code I wrote follows the first pattern instead of the later, but it is easy to change.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java (hunk 2 above)
Comment (gortiz) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java:109]: BTW2: Given that the returned value of `getAllIndexTypes` should be a `Set`, I had to use two attributes: a set that is returned by this method and a map that is used by `get` methods

Comment (saurabhd336) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java:109]: return `map.values()` instead?
Also, looks like the indexType's id is the unique identifier for the index type. Should we onverride equals and hashCode of the class to only include id?

The findAny() at the end of the filter seems a bit awkward :)

Comment (gortiz) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java:109]: > return map.values() instead?

map.values() is a collection, not a set.

> Also, looks like the indexType's id is the unique identifier for the index type. Should we onverride equals and hashCode of the class to only include id?

That is a complex question that may deserve a discussion, but that actually doesn't matter in a scenario where there are more than one implementation for a given index type id, what I'm calling index overriding. I don't know if this is something we expect to have soon, so I don't know if this is actually important.

The thing is, if we choose to define equality as "id is the same" and we have two different implementations in the classpath... the effects may be complex to analyze. Code in this branch doesn't actually use it that much, but other PRs will heavily use IndexType as hash map keys. 

As said, given that I don't think we are going to support index overriding, I guess we can add the equality check.

Comment (gortiz) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java:109]: If you don't mind, I'm going to add that in https://github.com/apache/pinot/pull/10184, given that there all indexes extend an abstract class, therefore we only need to add it once.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/dictionary/DictionaryIndexType.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+  @Override
+  public IndexCreator createIndexCreator(IndexCreationContext context, IndexConfig indexConfig)
+      throws Exception {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public IndexReaderFactory<IndexReader> getReaderFactory() {
+    throw new UnsupportedOperationException();
```
Comment (61yao) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/dictionary/DictionaryIndexType.java:74]: Some functions are unsupported. some are implemented. 

I assume the unsupported functions will be added in next PR. Is that the case?

Comment (gortiz) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/dictionary/DictionaryIndexType.java:74]: yep. They are implemented in https://github.com/apache/pinot/pull/10184

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/range/RangeIndexType.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.Schema;
+
+
+public class RangeIndexType implements IndexType<IndexConfig, IndexReader, IndexCreator> {
+
+  /**
+   * The default range index version used when not specified in the TableConfig.
+   *
+   * This value should be equal to the one used in {@link org.apache.pinot.spi.config.table.IndexingConfig}
```
Comment (klsince) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/range/RangeIndexType.java:45]: is this comment for this INSTANCE constant? I feel it's not quite related, but I may have missed some details.

Comment (gortiz) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/range/RangeIndexType.java:45]: Good catch and a funny story. I created this PR by using diff from changes in [`index-spi` branch](https://github.com/apache/pinot/pull/10184) and then removing things I didn't want to show in this PR. Here I removed a static constant but didn't removed the Javadoc.

The problem with my approach is that now if I delete the javadoc here, once I merge this with #10184 the comment will be removed without conflict and I may forget about it. Therefore what I'm going to do is to bring back the static constant. It is not going to be used in this PR, but it is going to ensure we don't accidentally remove the javadoc.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/StandardIndexes.java [hunk 6]
```diff
@@ -39,21 +39,64 @@
...
  *   providing a typesafe way to get these references. Instead of having to write something like
  *   {@code (IndexType<BloomFilterConfig, BloomFilterReader, IndexCreator>)
- *   IndexService.getInstance().getOrThrow("bloom_filter")},
+ *   IndexService.getInstance().get("bloom_filter")},
  *   a caller can simply use {@link StandardIndexes#bloomFilter()}
  * </p>
  */
 @SuppressWarnings("unchecked")
 public class StandardIndexes {
+  public static final String FORWARD_ID = "forward_index";
```
[the remark this answers is not in the document]
Comment (gortiz) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/StandardIndexes.java:48]: It was important because there were some code that iterate over the constants in a very specific order. In these cases I kept the order by explicitly introduce an order in the iteration, which is more verbose, but IMHO better as it explicitly introduce the order where the order really matters instead of implicitly use an order defined somewhere else.

But as you said, I had to deal with that in #10184. This is one of the reasons that PR is going to be trickier to merge than this one.

**Review by klsince — APPROVED** [review]

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/range/RangeIndexType.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+   */
+  public static final int DEFAULT_RANGE_INDEX_VERSION = 2;
```
Comment (61yao) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/range/RangeIndexType.java:47]: Not sure where this would be used.. and how we change the index version

**Review by 61yao — APPROVED** [review]

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/range/RangeIndexType.java (hunk 7 above)
Comment (gortiz) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/range/RangeIndexType.java:47]: It will be used in the next PR. We have been talking about it [here](https://github.com/apache/pinot/pull/10193#discussion_r1146922455). 

TL;DR: The const is not used here, but I added the javadoc by mistake on this PR. We can either add here the const (no problematic), let the javadoc written without the attribute (which seems rare) or remove here the javadoc. The later may seem to be the best option, but if I do that, the javadoc will be removed from the full PR once I merge the changes. If I don't remember to add it later, we will lose the javadoc. Given that this process is error prone, I decided to just bring the const to this PR, which causes no problem at the cost of having a const we don't actually use here.

>  and how we change the index version
This is the range index version we use right now.

**Review by KKcorps — APPROVED** [review]
